File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.java
class Solution {
    public long subArrayRanges(int[] arr) {
        int n=arr.length;
        int[] leftmin=new int[n];
        int[] leftmax=new int[n];
        int[] rightmin=new int[n];
        int[] rightmax=new int[n];
        Stack<Integer> st=new Stack<>();
        for (int i=0;i<n;i++){
            while(!st.isEmpty() && arr[st.peek()]>arr[i]) st.pop();
            leftmin[i]=st.isEmpty()?(i+1):i-st.peek();
            st.push(i);
        }
        while(!st.isEmpty()) st.pop();
        for (int i=n-1;i>=0;i--){
            while(!st.isEmpty() && arr[st.peek()]>=arr[i]) st.pop();
            rightmin[i]=st.isEmpty()?(n-i):st.peek()-i;
            st.push(i);
        }
        while(!st.isEmpty()) st.pop();
        for (int i=0;i<n;i++){
            while(!st.isEmpty() && arr[st.peek()]<arr[i]) st.pop();
            leftmax[i]=st.isEmpty()?(i+1):i-st.peek();
            st.push(i);
        }
        while(!st.isEmpty()) st.pop();
        for (int i=n-1;i>=0;i--){
            while(!st.isEmpty() && arr[st.peek()]<=arr[i]) st.pop();
            rightmax[i]=st.isEmpty()?(n-i):st.peek()-i;
            st.push(i);
        }
        while(!st.isEmpty()) st.pop();
        long result=0;
        for (int i=0;i<n;i++){
            long maxi=(long)arr[i]*leftmax[i]*rightmax[i];
            long mini=(long)arr[i]*leftmin[i]*rightmin[i];
            result+=maxi-mini;
        }
        return result;
    }
}

File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.cpp
class Solution {
public:
    long long subArrayRanges(vector<int>& arr) {
        int n=arr.size();
        vector<int> leftmin(n),leftmax(n),rightmin(n),rightmax(n);
        stack<int> st;
        for (int i=0;i<n;i++){
            while(!st.empty() && arr[st.top()]>arr[i]) st.pop();
            leftmin[i]=st.empty()?(i+1):i-st.top();
            st.push(i);
        }
        while(!st.empty()) st.pop();
        for (int i=n-1;i>=0;i--){
            while(!st.empty() && arr[st.top()]>=arr[i]) st.pop();
            rightmin[i]=st.empty()?(n-i):st.top()-i;
            st.push(i);
        }
        while(!st.empty()) st.pop();
        for (int i=0;i<n;i++){
            while(!st.empty() && arr[st.top()]<arr[i]) st.pop();
            leftmax[i]=st.empty()?(i+1):i-st.top();
            st.push(i);
        }
        while(!st.empty()) st.pop();
        for (int i=n-1;i>=0;i--){
            while(!st.empty() && arr[st.top()]<=arr[i]) st.pop();
            rightmax[i]=st.empty()?(n-i):st.top()-i;
            st.push(i);
        }
        while(!st.empty()) st.pop();
        long long result=0;
        for (int i=0;i<n;i++){
            long long maxi=(long long)arr[i]*leftmax[i]*rightmax[i];
            long long mini=(long long)arr[i]*leftmin[i]*rightmin[i];
            result+=maxi-mini;
        }
        return result;
    }
};